crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -0,0 +1,181 @@
+    let command = if std::mem::size_of::<T>() > 0 {
+        command.cast::<T>().read_unaligned()
+    } else {
+        // NOTE: This is necessary because if the `CommandQueueInner` is only filled with 0-sized
+        // commands the `bytes` vec will never allocate. Then means that `bytes.as_ptr()` could be null
+        // and reading a null pointer is always UB.
+        // However according to https://doc.rust-lang.org/std/ptr/index.html
+        // "The canonical way to obtain a pointer that is valid for zero-sized accesses is NonNull::dangling"
+        // therefore the below code is safe to do.
+        let ptr = std::ptr::NonNull::<T>::dangling().as_ptr();
+        ptr.cast::<T>().read()
+    };
+    command.write(world);
+}
+
+struct CommandMeta {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, world: &mut World),
+}
+
+#[derive(Default)]
+pub(crate) struct CommandQueueInner {
+    bytes: Vec<u8>,
+    metas: Vec<CommandMeta>,
+}
+
+// SAFE: All commands [`Command`] implement [`Send`]
+unsafe impl Send for CommandQueueInner {}
+
+// SAFE: `&CommandQueueInner` never gives access to the inner commands.
+unsafe impl Sync for CommandQueueInner {}
+
+impl CommandQueueInner {
+    /// Push a new `command` onto the queue.
+    #[inline]
+    pub fn push<C>(&mut self, command: C)
+    where
+        C: Command,
+    {
+        let size = std::mem::size_of::<C>();
+        let old_len = self.bytes.len();
+
+        self.metas.push(CommandMeta {
+            offset: old_len,
+            func: invoke_command::<C>,
+        });
+
+        if size > 0 {
+            self.bytes.reserve(size);
+
+            // SAFE: The internal `bytes` vector has enough storage for the
+            // command (see the call the `reserve` above), and the vector has
+            // its length set appropriately.
+            // Also `command` is forgotten at the end of this function so that
+            // when `apply` is called later, a double `drop` does not occur.
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    &command as *const C as *const u8,
+                    self.bytes.as_mut_ptr().add(old_len),
+                    size,
+                );
+                self.bytes.set_len(old_len + size);
+            }
+        }
+
+        std::mem::forget(command);
+    }
+
+    /// Invoke each command `func` for each inserted value with `world`
+    /// and then clears the internal bytes/metas command vectors.
+    pub fn apply(&mut self, world: &mut World) {
+        let byte_ptr = self.bytes.as_mut_ptr();
+        for meta in self.metas.iter() {
+            // The implementation of `invoke_command` is safe for the according Command type.
+            // The bytes are safely cast to their original type, safely read, and then dropped.
+            unsafe {
+                (meta.func)(byte_ptr.add(meta.offset), world);
+            }
+        }
+
+        self.bytes.clear();
```

Comment:
This isn't panic safe:

```rust
struct EvilCommand(String);
impl Command for EvilCommand {
 fn apply (){panic!()}
}

fn main(){
let mut commands = Commands::new();
commands.push(EvilCommand("Ahahahahah!".to_owned()));
let _ = std::panic::catch_unwind(|| commands.apply() );
commands.apply(); // Use after free/double free of the inner string
}
```

Reply:
You could use `self.metas.drain(..)` instead of `self.metas.iter()`. This will ensure that `self.metas` is empty in case of a panic.